### PR TITLE
test: stabilise TlsTest

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

fixes an issue where the TlsTest was failing because the file watcher wasn't detecting the change of the Tls cert from valid to invalid.  This was because of a limitation of JDK 8, which has only second resolution for `FileTime`, meaning the `WatchService` can miss updated made to files that existed before the watch was registered, if the update happened after the watch was started and the updated didn't change the last modified time, i.e. the last modified time both before and after the update was in the second second.

The fix is to insert a sleep to ensure at least one whole second has elapsed.

In production code, this one second window of bad behaviour is highly unlikely to be hit, and there's no real way to solve it, except upgrading to JDK 9.

### Testing done 

Usual

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:84]: why did we change this to swallow interrupted exceptions? seems like an atni-pattern

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:101]: should we backoff or something here? I feel like polling every second is going to make this thread take up way more resources that it needs to (how often do certs change? monthly?)

I'm also not entirely sure why this poll is necessary (why not just use `take`)?

**Review by agavra — COMMENTED** [review]
Nice sleuthing! Not sure I totally understand some parts though...

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java
Comment (vcrfxia) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java:133]: I've seen it take a number of seconds for the server to restart with the new cert after it's loaded. Even if the test fails, we should wait for the new cert to be loaded, which is why there's an `assertThatEventually` here, otherwise the next test that runs will likely fail.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/api/server/FileWatcherTest.java
Comment (vcrfxia) [inline ksqldb-rest-app/src/test/java/io/confluent/ksql/api/server/FileWatcherTest.java:107]: nit:
```suggestion
  public void shouldStopIfDirectoryDeleted() throws Exception {
```
Same for the test below.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (vcrfxia) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:101]: +1 I'm confused why this change was made.

**Review by vcrfxia — APPROVED** [review]
LGTM minus the comments inline. Thanks!

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:84]: Because the code doesn't interrupt the thread.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/api/server/FileWatcherTest.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/api/server/FileWatcherTest.java:107]: Well... I don't think it _should_, it just happens that it _will_.  Hence the naming...

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:101]: `take` blocks and requires the use of interrupts to shutdown the thread.  This is generally considered an anti-pattern and goes against the "use exceptions for exceptional cases only" paradigm. 

Exiting the poll every second,,and then immediately looping and re-entering the poll, will take almost no CPU cycles and allows it to exit gracefully.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java:133]: Actually, as I've changed `ServerKeyStore` to be non-static, there's no need for this to happen any more.  There's only one test in the class, and this class won't interfere with other classes... so no need to change, right?

Comment (vcrfxia) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java:133]: Looks like my comment from the previous PR wasn't addressed before merging: https://github.com/confluentinc/ksql/pull/6053#discussion_r474003954

I contend there _should_ be multiple tests running in this class 😄

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (agavra) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:101]: > This is generally considered an anti-pattern and goes against the "use exceptions for exceptional cases only" paradigm.

I agree that requiring interrupts to shutdown the thread should be avoided, but I've seen different workarounds to this problem (specifically when using `BlockingQueue`) - the most popular of which is adding a sentinel into the queue to wake up the blocked thread.

I'm not familiar with `WatchService`, but is it possible to do something similar (e.g. touch a lock file)? I understand your reasoning now, so if there isn't clean we can do then I'm happy proceeding with this

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (agavra) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:84]: I still don't think that we should be swallowing the interrupted exception - I don't see the benefit (we lose nothing by keeping it, we might lose something by removing it). Just because the code today doesn't intentionally interrupt this thread, it doesn't mean that future code won't or that there is some code we didn't audit that does.

File: ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/test/java/io/confluent/ksql/api/TlsTest.java:133]: I'll sort that and this then ;)

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:84]: It's generally considered bad practice to us interrupts to control thread lifecycles.  Given we're not, if someone wants to shutdown this thread they should call `shutdown` on the watcher and the code will do the right thing.  If they try interrupting the thread then the code will log a `unexpected exception` error and continue - this to me is the _right_ behaviour: interrupt exceptions aren't expected and shouldn't be used to control the lifecycle of this thread.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:101]: Sounds like more code for the sake of it.   Are you really worried about a thread waking every second and looping?

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java
Comment (agavra) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/api/server/FileWatcher.java:84]: I didn't mean to suggest that we should shutdown the thread on interrupt but at we should catch it and reset the thread's interrupt flag to make sure we don't loose this information. I'll defer to Brian Goetz's article as he makes the point way better than I can: https://www.ibm.com/developerworks/library/j-jtp05236/index.html

The relevant part:
> If you catch InterruptedException but cannot rethrow it, you should preserve evidence that the interruption occurred so that code higher up on the call stack can learn of the interruption and respond to it if it wants to. This task is accomplished by calling interrupt() to "reinterrupt" the current thread, as shown in Listing 3. At the very least, whenever you catch InterruptedException and don't rethrow it, reinterrupt the current thread before returning. 
>
> The worst thing you can do with InterruptedException is swallow it -- catch it and neither rethrow it nor reassert the thread's interrupted status. The standard approach to dealing with an exception you didn't plan for -- catch it and log it -- also counts as swallowing the interruption because code higher up on the call stack won't be able to find out about it. (Logging InterruptedException is also just silly because by the time a human reads the log, it is too late to do anything about it.)

**Pull request merged**
